### PR TITLE
add node2vec baseline

### PR DESCRIPTION
**Pull request by yiweny**
We directly use link prediction train table as data source since the links between left hand side and right hand side nodes already form a graph, we can run node2vec model on it and get the embedding of each node in the graph then similarly, we can assign a score on each links based on cosine similarity.
No features are used.

### CODE REVIEW

File: examples/node2vec_link.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+parser.add_argument("--embedding_dim", type=int, default=128)
+parser.add_argument("--walk_length", type=int, default=10)
+parser.add_argument("--context_size", type=int, default=5)
+parser.add_argument("--num_workers", type=int, default=1)
+parser.add_argument("--log_dir", type=str, default="results")
+args = parser.parse_args()
+
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+if torch.cuda.is_available():
+    torch.set_num_threads(1)
```
Comment (rusty1s) [inline examples/node2vec_link.py:34]: why?

File: examples/node2vec_link.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+assert task.task_type == TaskType.LINK_PREDICTION
+
+col_to_stype_dict = dataset2inferred_stypes[args.dataset]
+
+data, col_stats_dict = make_pkey_fkey_graph(
+    dataset.db,
+    col_to_stype_dict=col_to_stype_dict,
+    text_embedder_cfg=TextEmbedderConfig(
+        text_embedder=GloveTextEmbedding(device=device), batch_size=256
+    ),
```
Comment (rusty1s) [inline examples/node2vec_link.py:52]: skip, no need to use text embedding.

File: examples/node2vec_link.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    cache_dir=os.path.join(root_dir, f"{args.dataset}_materialized_cache"),
+)
+
+
+num_src_nodes = task.num_src_nodes
+df = task.train_table.df.explode(task.dst_entity_col)
+edge = torch.stack(
```
Comment (rusty1s) [inline examples/node2vec_link.py:59]: Let's call it `edge_index`.

File: examples/node2vec_link.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    [
+        torch.from_numpy(df[task.src_entity_col].astype(int).values),
+        (torch.from_numpy(df[task.dst_entity_col].astype(int).values + num_src_nodes)),
+    ]
```
Comment (rusty1s) [inline examples/node2vec_link.py:63]: You are missing the reverse direction here! This is crucial, e.g., via `torch_geometric.utils.to_undirected()`.

File: examples/node2vec_link.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    cache_dir=os.path.join(root_dir, f"{args.dataset}_materialized_cache"),
+)
+
+
+num_src_nodes = task.num_src_nodes
+df = task.train_table.df.explode(task.dst_entity_col)
+edge = torch.stack(
+    [
+        torch.from_numpy(df[task.src_entity_col].astype(int).values),
+        (torch.from_numpy(df[task.dst_entity_col].astype(int).values + num_src_nodes)),
```
Comment (rusty1s) [inline examples/node2vec_link.py:62]: Would be good to only use training edges here.

Comment (yiweny) [reply examples/node2vec_link.py:62]: The `df` here is 
```
df = task.train_table.df.explode(task.dst_entity_col)
```
So I assume it's only using the training edges?

File: examples/node2vec_link.py (hunk 1 above)
Comment (yiweny) [reply examples/node2vec_link.py:34]: I was copying this from `gnn_node.py`. Removed here and in the `gnn_node.py` file.

File: examples/node2vec_link.py (hunk 2 above)
Comment (yiweny) [reply examples/node2vec_link.py]: Actually if the data contains a text column but this is not provided, pytorch-frame will report an error.

File: examples/gnn_node.py [hunk 6]
```diff
@@ -37,8 +37,7 @@
 args = parser.parse_args()
 
 device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
-if torch.cuda.is_available():
-    torch.set_num_threads(1)
```
Comment (weihua916) [inline examples/gnn_node.py:41]: Don't delete

File: examples/node2vec_link.py [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+    # Might be good to filter out entities based on time col in the future
+    return dst_entity_table.df[task.dst_entity_col].values + num_src_nodes
+
+
+@torch.no_grad()
+def test(table: Table, k):
+    model.eval()
+    z = model()
+    lhs = z[table.df[task.src_entity_col].values]
+    rhs = z[get_valid_test_dst_entities()]
```
Comment (weihua916) [inline examples/node2vec_link.py:112]: rename all
```
lhs -> src
rhs -> dst
```

File: examples/node2vec_link.py (hunk 1 above)
Comment (rishabh-ranjan) [reply examples/node2vec_link.py]: actually, this was important to run the script on many datasets and tasks in parallel (otherwise threading causes it to be cpu bottlenecked). Weihua suggested to have this in the public scripts as well, and I have checked that it does not hurt runtime significantly.

Comment (rishabh-ranjan) [reply examples/node2vec_link.py]: @yiweny please add back

File: examples/node2vec_link.py [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+
+for epoch in range(1, args.epochs + 1):
+    train_loss = train()
+    val_pred = test(task.val_table, task.eval_k)
+    val_metrics = task.evaluate(val_pred, task.val_table)
+    print(
+        f"Epoch: {epoch:02d}, Train loss: {train_loss}, " f"Val metrics: {val_metrics}"
+    )
+
+    if val_metrics[tune_metric] >= best_val_metric:
```
Comment (weihua916) [inline examples/node2vec_link.py:133]: Use `>` instead of `>=`

File: examples/node2vec_link.py [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+    for col, stype in table_col_to_stype.items():
+        if stype == torch_frame.text_embedded:
+            col_to_stype_dict[table_name][col] = torch_frame.categorical
+
+
+data, col_stats_dict = make_pkey_fkey_graph(
+    dataset.db,
+    col_to_stype_dict=col_to_stype_dict,
+    cache_dir=os.path.join(root_dir, f"{args.dataset}_materialized_cache"),
+)
```
Comment (weihua916) [inline examples/node2vec_link.py:59]: These are not used, so we can just delete these lines.

File: examples/node2vec_link.py [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+
+
+num_src_nodes = task.num_src_nodes
+df = task.train_table.df.explode(task.dst_entity_col)
+# edge from src to dst
+edge_index = torch.stack(
+    [
+        torch.from_numpy(df[task.src_entity_col].astype(int).values),
+        (torch.from_numpy(df[task.dst_entity_col].astype(int).values + num_src_nodes)),
```
Comment (weihua916) [inline examples/node2vec_link.py:68]: Can you add a comment on what you are doing here?

File: examples/node2vec_link.py [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+    return dst_entity_table.df[task.dst_entity_col].values + num_src_nodes
+
+
+@torch.no_grad()
+def test(table: Table, k):
+    model.eval()
+    z = model()
+    src = z[table.df[task.src_entity_col].values]
+    dst = z[get_valid_test_dst_entities()]
+    _, indices = torch.topk(src @ dst.T, k)
```
Comment (weihua916) [inline examples/node2vec_link.py:114]: Node2vec is optimized to predict random walk probability. It's not explicitly optimizing inner-product-based prediction accuracy. It's a bit odd to directly use node2vec's embeddings to make prediction.

Instead, what people usually do is to pre-train node2vec embeddings and then train another classifier on top of the node2vec embeddings to make the final prediction. @rusty1s can give more details on implementation.

**Review by weihua916 — COMMENTED** [review]
Gave some comments.

My major comment is node2vec in the current implementation seems like a very weak baseline, since it's using node2vec embeddings directly for final prediction (via inner product).

How does it actually perform on the relbench link prediction tasks?

File: examples/node2vec_link.py (hunk 9 above)
Comment (yiweny) [reply examples/node2vec_link.py:59]: Thanks!

**Comment by rishabh-ranjan** [comment]
@yiweny what is the status of this PR? Can we close it? Can we delete the branch or would you like to keep it?

**Comment by weihua916** [comment]
Let's close this PR for now.
